### PR TITLE
FOUNDATIONS: Document CreateDirectorySasToken

### PR DESCRIPTION
**Pull request by MarcusEnrique**

### CODE REVIEW

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Logic.CreateDirectorySasToken.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            DateTimeOffset futureDateTimeOffset = GetRandomFutureDateTimeOffset();
+            string expectedSasToken = GetRandomString();
+
+            this.blobStorageBrokerMock.Setup(broker =>
+                broker.CreateDirectorySasTokenAsync(
+                    randomContainer,
+                    directoryPath,
+                    accessPolicyIdentifier,
+                    futureDateTimeOffset))
+                .ReturnsAsync(expectedSasToken);
```
[the remark this answers is not in the document]
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Logic.CreateDirectorySasToken.cs:31]: the `.ReturnsAsync` method call should be indented relative to the arguments

File: ISL.ReIdentification.Core/Services/Foundations/Documents/IDocumentService.cs [hunk 2]
```diff
@@ -20,5 +20,7 @@ public interface IDocumentService
         ValueTask AddContainerAsync (string container);
         ValueTask AddFolderAsync(string container, string folder);
         ValueTask<string> GetDownloadLinkAsync(string fileName, string container, DateTimeOffset expiresOn);
+        ValueTask<string> CreateDirectorySasTokenAsync(
```
Comment (MichaelOSullivanAnswer) [inline ISL.ReIdentification.Core/Services/Foundations/Documents/IDocumentService.cs:23]: new line before multi-line piece of code
```suggestion

        ValueTask<string> CreateDirectorySasTokenAsync(
```

Comment (MarcusEnrique) [reply ISL.ReIdentification.Core/Services/Foundations/Documents/IDocumentService.cs:23]: Done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Validations.CreateDirectorySasToken.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                key: "accessPolicyIdentifier",
+                values: "Text is invalid");
+
+            var expectedDocumentValidationException = new DocumentValidationException(
+                message: "Document validation error occurred, please fix errors and try again.",
+                innerException: invalidDocumentException);
+
+            // when
+            ValueTask<string> createDirectorySasTokenTask =
+                this.documentService.CreateDirectorySasTokenAsync(invalidContainer, invalidDirectoryPath, invalidAccessPolicyIdentifier, invalidDateTimeOffset);
```
[the remark this answers is not in the document]
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Validations.CreateDirectorySasToken.cs:50]: Done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Logic.CreateDirectorySasToken.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Logic.CreateDirectorySasToken.cs:31]: Done

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs [hunk 4]
```diff
@@ -93,5 +93,15 @@ public ValueTask<string> GetDownloadLinkAsync(string fileName, string container,
             
             return await this.blobStorageBroker.GetDownloadLinkAsync(fileName, container, expiresOn);
         });
+
+        public ValueTask<string> CreateDirectorySasTokenAsync(
+            string container, string directoryPath, string accessPolicyIdentifier, DateTimeOffset expiresOn) =>
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs:98]: Wrap properties vertically

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            string someAccessPolicyIdentifier = GetRandomString();
+            DateTimeOffset someDateTimeOffset = GetRandomFutureDateTimeOffset();
+
+            var expectedDependencyValidationException = new DocumentDependencyValidationException(
+                message: "Document dependency validation error occurred, please fix errors and try again.",
+                innerException: dependencyValidationException);
+
+            this.blobStorageBrokerMock.Setup(broker =>
+                broker.CreateDirectorySasTokenAsync(
+                    someContainer, someDirectoryPath, someAccessPolicyIdentifier, someDateTimeOffset))
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:35]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                        .ThrowsAsync(dependencyValidationException);
+
+            // when
+            ValueTask<string> createDirectorySasTask =
+                this.documentService.CreateDirectorySasTokenAsync(
+                    someContainer, someDirectoryPath, someAccessPolicyIdentifier, someDateTimeOffset);
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:41]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+            DocumentDependencyValidationException actualDocumentServiceException =
+                await Assert.ThrowsAsync<DocumentDependencyValidationException>(createDirectorySasTask.AsTask);
+
+            // then
+            actualDocumentServiceException.Should().BeEquivalentTo(expectedDependencyValidationException);
+
+            this.blobStorageBrokerMock.Verify(broker =>
+                broker.CreateDirectorySasTokenAsync(
+                    someContainer, someDirectoryPath, someAccessPolicyIdentifier, someDateTimeOffset),
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:51]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            string someAccessPolicyIdentifier = GetRandomString();
+            DateTimeOffset someDateTimeOffset = GetRandomFutureDateTimeOffset();
+
+            var expectedDocumentDependencyException = new DocumentDependencyException(
+                message: "Document dependency error occurred, please fix errors and try again.",
+                innerException: dependencyException);
+
+            this.blobStorageBrokerMock.Setup(broker =>
+                broker.CreateDirectorySasTokenAsync(
+                    someContainer, someDirectoryPath, someAccessPolicyIdentifier, someDateTimeOffset))
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:79]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+                        .ThrowsAsync(dependencyException);
+
+            // when
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:82]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            ValueTask<string> createDirectorySasTokenTask =
+                this.documentService.CreateDirectorySasTokenAsync(
+                    someContainer, someDirectoryPath, someAccessPolicyIdentifier, someDateTimeOffset);
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:85]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+                message: "Failed service document error occurred, contact support.",
+                innerException: someException);
+
+            var expectedDocumentServiceException = new DocumentServiceException(
+                message: "Document service error occurred, contact support.",
+                innerException: failedServiceDocumentException);
+
+            this.blobStorageBrokerMock.Setup(broker =>
+                broker.CreateDirectorySasTokenAsync(
+                    someContainer, someDirectoryPath, someAccessPolicyIdentifier, someDateTimeOffset))
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:127]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+            DocumentServiceException actualDocumentServiceException =
+                await Assert.ThrowsAsync<DocumentServiceException>(createDirectorySasTokenTask.AsTask);
+
+            // then
+            actualDocumentServiceException.Should().BeEquivalentTo(expectedDocumentServiceException);
+
+            this.blobStorageBrokerMock.Verify(broker =>
+                broker.CreateDirectorySasTokenAsync(
+                    someContainer, someDirectoryPath, someAccessPolicyIdentifier, someDateTimeOffset),
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:143]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Validations.CreateDirectorySasToken.cs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+                values: "Text is invalid");
+
+            var expectedDocumentValidationException = new DocumentValidationException(
+                message: "Document validation error occurred, please fix errors and try again.",
+                innerException: invalidDocumentException);
+
+            // when
+            ValueTask<string> createDirectorySasTokenTask =
+                this.documentService.CreateDirectorySasTokenAsync(
+                    invalidContainer, invalidDirectoryPath, invalidAccessPolicyIdentifier, invalidDateTimeOffset);
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Validations.CreateDirectorySasToken.cs:51]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Validations.CreateDirectorySasToken.cs [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+            DocumentValidationException actualDocumentValidationException =
+                await Assert.ThrowsAsync<DocumentValidationException>(createDirectorySasTokenTask.AsTask);
+
+            // then
+            actualDocumentValidationException.Should().BeEquivalentTo(expectedDocumentValidationException);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogErrorAsync(It.Is(SameExceptionAs(
+                    expectedDocumentValidationException))),
+                        Times.Once);
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Validations.CreateDirectorySasToken.cs:62]:
```suggestion
            this.loggingBrokerMock.Verify(broker =>
                broker.LogErrorAsync(It.Is(SameExceptionAs(expectedDocumentValidationException))),
                    Times.Once);
```

File: ISL.ReIdentification.Core/Services/Foundations/Documents/IDocumentService.cs (hunk 2 above)
Comment (cjdutoit) [reply ISL.ReIdentification.Core/Services/Foundations/Documents/IDocumentService.cs:24]: fix

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs (hunk 5 above)
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:35]: done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs (hunk 6 above)
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:41]: done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs (hunk 7 above)
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:51]: done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs (hunk 8 above)
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:79]: done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs (hunk 9 above)
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:82]: done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs (hunk 10 above)
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:85]: done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+            DocumentDependencyException actualDocumentDependencyException =
+                await Assert.ThrowsAsync<DocumentDependencyException>(createDirectorySasTokenTask.AsTask);
+
+            // then
+            actualDocumentDependencyException.Should().BeEquivalentTo(expectedDocumentDependencyException);
+
+            this.blobStorageBrokerMock.Verify(broker =>
+                broker.CreateDirectorySasTokenAsync(
+                    someContainer, someDirectoryPath, someAccessPolicyIdentifier, someDateTimeOffset),
```
[the remark this answers is not in the document]
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:95]: done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs (hunk 12 above)
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Exceptions.CreateDirectorySasToken.cs:143]: done

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Validations.CreateDirectorySasToken.cs (hunk 13 above)
Comment (MarcusEnrique) [reply ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.Validations.CreateDirectorySasToken.cs:51]: done

**Review by cjdutoit — APPROVED** [review]

**Pull request merged**